File: 5-1-2.cpp
#include<bits/stdc++.h>
using namespace std;
typedef long long ll;
void do_add_query(int psum[], int i, int j, int k, int n) {
    psum[i] += k;
    if (j + 1 < n) psum[j + 1] -= k;
}
ll get_sum(int A[], int psum[], int i, int j, int n) {
    for (int t = 1; t < n; t++) psum[t] += psum[t - 1];
    ll ret = 0;
    for (int t = i; t <= j; t++) ret += psum[t] + A[t];
    return ret;
}
int main() {
    ios::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);

    int n, m; cin >> n >> m;
    int A[100004];
    for (int i = 0; i < n; i++) cin >> A[i];
    int psum[100004] = { 0 };
    while (m--) {
        int op, i, j, k; cin >> op;
        if (op == 1) {
            cin >> i >> j >> k;
            do_add_query(psum, i, j, k, n);
        }
        else {
            cin >> i >> j;
            cout << get_sum(A, psum, i, j, n) << '\n';
        }
    }
}

File: 5-1-2.java
import java.util.Scanner;

public class Main {
    public static void do_add_query(int[] psum, int i, int j, int k, int n) {
        psum[i] += k;
        if (j + 1 < n) psum[j + 1] -= k;
    }

    public static long get_sum(int[] A, int[] psum, int i, int j, int n) {
        for (int t = 1; t < n; t++) psum[t] += psum[t - 1];
        long ret = 0;
        for (int t = i; t <= j; t++) ret += psum[t] + A[t];
        return ret;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        int m = scanner.nextInt();
        int[] A = new int[100004];
        for (int i = 0; i < n; i++) A[i] = scanner.nextInt();
        int[] psum = new int[100004];
        while (m-- > 0) {
            int op = scanner.nextInt();
            if (op == 1) {
                int i = scanner.nextInt();
                int j = scanner.nextInt();
                int k = scanner.nextInt();
                do_add_query(psum, i, j, k, n);
            } else {
                int i = scanner.nextInt();
                int j = scanner.nextInt();
                System.out.println(get_sum(A, psum, i, j, n));
            }
        }
    }
}
